### PR TITLE
pipeline-service

### PR DESCRIPTION
**Pull request by malaffoon**
@gregcharles, there is (at least) one more thing that needs to be done here: set up the exam-processor with the system script for validation. I'm going to add a test that asserts that exam-processor has a non-null pipeline, but obviously i'll have to @Ignore that until you make that change.

### CODE REVIEW

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * (and the cache is tenant-aware).
+ */
+public class DefaultPipelineService implements PipelineService {
+    private static final Logger logger = LoggerFactory.getLogger(DefaultPipelineService.class);
+
+    private final TenantIdResolver tenantIdResolver;
+    private final PipelineFactory pipelineFactory;
+
+    private final LoadingCache<String, Optional<String>> versions;  // tenantId_pipelineCode -> active version
+    private final Map<String, Optional<Pipeline>> pipelines;        // tenantId_pipelineCode -> pipeline
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java]: We want to cache the active version with a short TTL, so we can detect publish changes. We want to cache the pipeline forever (well, until the active version changes).

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        versions = CacheBuilder.newBuilder()
+                .maximumSize(100)
+                .expireAfterWrite(5, TimeUnit.MINUTES)
+                .build(new CacheLoader<String, Optional<String>>() {
+                    @Override
+                    public Optional<String> load(final String key) {
+                        return Optional.ofNullable(pipelineRepository.findActiveVersion(key.split("_")[1]));
+                    }
+                });
```
Comment (esotrope) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java]: i think you can externalize everything up to the build method with a method called "fromSpec" or something.

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/ExamProcessorValidationTest.java [hunk 3]
```diff
@@ -251,6 +221,37 @@ private void setReposToSunnyDay() {
...
+        // the pipeline has just the system post script
+
+        // this won't get called
+        final PublishedPipelineRepository repository = mock(PublishedPipelineRepository.class);
+
+        // to avoid interfering with other tests we have to have a custom script source
+        final ScriptSource scriptSource = new TestScriptSource();
+
+        // we need a system script property resolver
+        final PropertyResolver propertyResolver = new SystemScriptPropertyResolver(
```
Comment (esotrope) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/ExamProcessorValidationTest.java]: wow big constructor x J

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java (hunk 2 above)
Comment (malaffoon) [reply common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java]: Oh. Yeah. That looks handy.

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java (hunk 1 above)
Comment (gregcharles) [reply common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java]: Makes sense, but you may be burying the lede here. Why did you pull this out of RDW-Common? Is that temporary -- or are the needs of ingest and reporting irreconcilable?

Comment (malaffoon) [reply common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java]: I think they are different needs. Ingest needs to cache versions because it is receiving many requests a second, and it only needs the active pipelines. Reporting wants to know everything about all pipelines and is in user-time, so caching is not good. I suppose the system would be more responsive if we cached pipelines for running tests and stuff, but that caching would be for any/all pipelines (inc. the development pipeline), not just the active published one.

Comment (gregcharles) [reply common/src/main/java/org/opentestsystem/rdw/ingest/common/script/DefaultPipelineService.java]: Oh, wait, I got confused with names. This isn't the factory. :)

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/ExamProcessorValidationTest.java (hunk 3 above)
Comment (gregcharles) [reply exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/ExamProcessorValidationTest.java]: Hey now! The property resolver is autowired with the repos needed to do the validations, plus a few more that seemed reasonable for future validations. There are just a lot of repos.

Comment (esotrope) [reply exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/ExamProcessorValidationTest.java]: haha. constructors like this are why dependency injection was invented. long live DI x D
